examples/cpp/reflection/proto_reflection_descriptor_database.cc

```diff
@@ -0,0 +1,183 @@
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#include "proto_reflection_descriptor_database.h"
+
+#include <vector>
+
+#include <grpc/support/log.h>
+
+namespace grpc {
+
+ProtoReflectionDescriptorDatabase::ProtoReflectionDescriptorDatabase(
+    std::unique_ptr<reflection::v1::ServerReflection::Stub> stub)
+    : stub_(std::move(stub)) {}
+
+ProtoReflectionDescriptorDatabase::~ProtoReflectionDescriptorDatabase() {}
+
+bool ProtoReflectionDescriptorDatabase::FindFileByName(
+    const string& filename, google::protobuf::FileDescriptorProto* output) {
+  if (cached_db_.FindFileByName(filename, output)) {
+    return true;
+  }
+
+  if (known_files_.find(filename) != known_files_.end()) {
+    return false;
+  }
+
+  ClientContext ctx;
+  reflection::v1::FileNameRequest request;
+  request.set_filename(filename);
+  reflection::v1::FileDescriptorProtoResponse response;
+
+  Status status = stub_->GetFileByName(&ctx, request, &response);
+  if (status.error_code() == StatusCode::OK) {
+    const google::protobuf::FileDescriptorProto* file_proto =
+        response.mutable_file_descriptor_proto();
+    known_files_.insert(file_proto->name());
+    cached_db_.Add(*file_proto);
+  } else if (status.error_code() == StatusCode::NOT_FOUND) {
+    gpr_log(GPR_INFO, "NOT_FOUND from server for FindFileByName()");
+  } else {
+    gpr_log(GPR_INFO, "Error on FindFileByName()");
+  }
+
+  return cached_db_.FindFileByName(filename, output);
+}
+
+bool ProtoReflectionDescriptorDatabase::FindFileContainingSymbol(
+    const string& symbol_name, google::protobuf::FileDescriptorProto* output) {
+  if (cached_db_.FindFileContainingSymbol(symbol_name, output)) {
+    return true;
+  }
+
+  if (missing_symbols_.find(symbol_name) != missing_symbols_.end()) {
+    return false;
+  }
+
+  ClientContext ctx;
+  reflection::v1::SymbolRequest request;
+  request.set_symbol(symbol_name);
+  reflection::v1::FileDescriptorProtoResponse response;
+
+  Status status = stub_->GetFileContainingSymbol(&ctx, request, &response);
+  if (status.error_code() == StatusCode::OK) {
+    const google::protobuf::FileDescriptorProto* file_proto =
+        response.mutable_file_descriptor_proto();
+    if (known_files_.find(file_proto->name()) == known_files_.end()) {
+      known_files_.insert(file_proto->name());
+      cached_db_.Add(*file_proto);
+    }
+  } else if (status.error_code() == StatusCode::NOT_FOUND) {
+    missing_symbols_.insert(symbol_name);
+    gpr_log(GPR_INFO, "NOT_FOUND from server for FindFileContainingSymbol()");
+  } else {
+    gpr_log(GPR_INFO, "Error on FindFileContainingSymbol()");
```

Comment:
We need to print out the error code and error message. We may hit this path often, for example when the server does not have reflection service.